File: 0001-two-sum/0001-two-sum.java
import java.util.HashMap;
import java.util.Map;

class Solution {
    public int[] twoSum(int[] nums, int target) {
        Map<Integer, Integer> seen = new HashMap<>(nums.length); // value → index

        for (int i = 0; i < nums.length; i++) {
            int complement = target - nums[i];
            Integer j = seen.get(complement);
            if (j != null) {          // match found
                return new int[] { j, i };
            }
            seen.put(nums[i], i);     // store after the check
        }
        return new int[0];            // never reached per problem guarantee
    }
}

File: 0001-two-sum/0001-two-sum.cpp
#include <vector>
#include <unordered_map>

class Solution {
public:
    std::vector<int> twoSum(const std::vector<int>& nums, int target) {
        std::unordered_map<int, int> seen;           
        seen.reserve(nums.size());                   
        
        for (int i = 0; i < static_cast<int>(nums.size()); ++i) {
            int complement = target - nums[i];
            auto it = seen.find(complement);
            if (it != seen.end()) {                  
                return {it->second, i};
            }
            seen[nums[i]] = i;    
                          
        }
        return {};
    }
};
